### DISCUSSION TITLE
How do I seed initial schema/data for self-hosted docker-compose setup?

### DISCUSSION THREAD

**Discussion by benrbowers** [supabase/supabase#12554] [Questions]
When I startup supabase with the cli (`supabase start`), migrations are run from supabase/migrations. How do I achieve similar when starting from docker compose?

Here's the docker-compose.yml I'm using: https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml

I cloned the repo using these instructions: https://supabase.com/docs/guides/self-hosting/docker

I see `docker/dev/data.sql` and `docker/volumes/db/init/data.sql`
Do I need to use those files somehow?

**Comment by SteveChurch** [comment]
This may or may not be of help depending on how or where you are self-hosting. 

https://github.com/digitalocean/supabase-on-do/discussions/15

But I now handle this as part of a cron which funs on my Droplet which is running the Supabase Self Hosted docker images.

**Comment by Vanuan** [comment]
I think it's either `supabase db reset --db-url=$DSN` or `supabase db push --db-url=$DSN`.

The `reset` command drops database and re-applies migrations and seed data. I think you don't want to erase your self-hosted database, do you?

The `push` commands the applies only the migrations missing in the database and doesn't apply seed data. This should be used like a normal deploy when you tested your app to be working with new migrations applied.

Be aware that your app's `supabase/migrations`  are not the only migrations applied. Additional schema and data is embedded inside the `supabase/postgres` docker image: https://github.com/supabase/postgres/tree/develop/migrations which are run when you initialise your `pg` container volume. There's probably a separate upgrade process when you upgrade the supabase images versions of your self-hosted setup.

The `docker/dev/data.sql` is probably an example file you can mount instead of your own `seed.sql` (or migrations). It looks to be an example described in the documentation here: https://supabase.com/docs/guides/auth/managing-user-data#creating-user-tables

Additionally, you can always use `supabase db dump` and `psql --file dump.sql` to work with schemas (and data) directly. Or use completely independent tools, like [Atlas](https://atlasgo.io/).

To summarise, the playbook when using the supabase migrations framework would look like this:

1. Create your new migrations (`supabase migration new` or `supabase db remote commit` or `supabase db diff`).
2. Test your new migrations with local environment (`supabase start` or `supabases db push` or `supabase db reset`)
3. Change the app to be compatible with both migration versions (use conditional migration version/names if needed).
4. Deploy and test the app to staging and production.
5. Deploy and test migrations in staging and production (`supabase db push`).
6. Remove conditionals from the app when everything works and all migrations are applied. Unless you want to revert migrations in the future.

If you use a tool like Atlas, you probably don't even need the Supabase CLI for db/migration management or the `supabase/migrations` files.

**Comment by benrbowers** [comment]
I've decided not to use self-hosting for now because I think I'll be able to coast on the free/$20 tiers of Supabase's managed hosting for quite some time.  So for now supabase's provided migration tools are sufficient. If I move to self-hosting in the future, I think Atlas (atlasgo.io, not MongoDB's Atlas) or dbmate look like solid options.

**Comment by mavcook** [comment]
Just in case others come across this, hopefully this saves you 3 minutes of continued searching.
Currently (supabase CLI v2.19.7) you can run the command below to seed your self-hosted database from your local dev machine:

```
supabase db push --include-seed --db-url \
    'postgres://postgres.stacysmom:fountains-of-wayne_can_hang_around_by_the_pool@supabat.bruceof.wayne.com:6543/gotitgoinon'
```

The `--db-url` is composed from the environment variables found in the [docker .env file](https://github.com/supabase/supabase/blob/9793cf73ca2916e1051612ca3fd599a9e8a75848/docker/.env.example#L6). You are of course familiar with, or are already using the industry-standard values, but just in case you need to copy them:
```
POOLER_TENANT_ID=stacysmom
POSTGRES_PASSWORD=fountains-of-wayne_can_hang_around_by_the_pool
SITE_URL=https://supabat.bruceof.wayne.com
POSTGRES_DB=gotitgoinon
```


Links to the relevant docs for connecting to your remote postgres are [here](https://supabase.com/docs/guides/self-hosting/docker#accessing-postgres) ([permalink](https://github.com/supabase/supabase/blob/9793cf73ca2916e1051612ca3fd599a9e8a75848/apps/docs/content/guides/self-hosting/docker.mdx?plain=1#L134-L142)).

**Reply by nicetomytyuk** [reply]
Is there a way to use a different file for seeding instead seed.sql? Because in my local environment my seed.sql adds lot of test data, while I have another file like init.sql where I have all static configuration data that must be import when the self-hosted is mounted.

**Reply by mavcook** [reply]
Yes I believe so. The `supabase db push` command looks at what you have in config.toml->[db.seed]->sql_paths, so you can temporarily put [whatever paths](https://github.com/supabase/supabase/blob/c4316c3ace60ff08352bbffad35cb2133eec9aa8/apps/docs/content/guides/local-development/seeding-your-database.mdx#splitting-up-your-seed-file) you need for the push. just add the `--dry-run` flag to the command to make sure it will indeed use the seed files you expect.

**Comment by nk9** [comment]
To answer the original question succinctly: you can put any initalization SQL in `./volumes/db/init/data.sql` and then mount it into the container using this in the `docker-compose.yml` under `services.db.volumes`:
```
      - ./volumes/db/init/data.sql:/docker-entrypoint-initdb.d/data.sql:Z
```

This will run any code you put there ONLY on first init of the database.
